components/org.wso2.identity.event.http.publisher/src/main/java/org/wso2/identity/event/http/publisher/internal/service/impl/HTTPEventPublisherImpl.java

```diff
@@ -84,72 +96,105 @@ public boolean canHandleEvent(EventContext eventContext) throws EventPublisherEx
-            String secret = webhook.getSecret();
-            sendWithRetries(eventPayload, eventContext, url, secret,
+            final String url = webhook.getEndpoint();
+            final String secret = webhook.getSecret();
+
+            sendWithRetries(eventProfileName, eventProfileUri, events,
+                    bodyJson, correlationId, tenantDomain, url, secret,
                     HTTPAdapterDataHolder.getInstance().getClientManager().getMaxRetries());
         }
     }
 
-    private void sendWithRetries(SecurityEventTokenPayload eventPayload, EventContext eventContext,
-                                 String url, String secret, int retriesLeft) {
+    /**
+     * Retries always reuse the frozen snapshots; no shared mutable state is read here.
+     */
+    private void sendWithRetries(String eventProfileName, String eventProfileUri, String events, String bodyJson,
+                                 String correlationId, String tenantDomain, String url, String secret,
+                                 int retriesLeft) {
 
         ClientManager clientManager = HTTPAdapterDataHolder.getInstance().getClientManager();
+
         final HttpPost request;
         try {
-            request = clientManager.createHttpPost(url, eventPayload, secret);
+            request = clientManager.createHttpPost(url, bodyJson, secret);
         } catch (HTTPAdapterException e) {
-            printPublisherDiagnosticLog(eventContext, eventPayload, url,
+            printPublisherDiagnosticLog(eventProfileName, eventProfileUri, events, url,
                     HTTPAdapterConstants.LogConstants.ActionIDs.PUBLISH_EVENT, DiagnosticLog.ResultStatus.FAILED,
                     "Failed to construct HTTP request for HTTP adapter publish.");
             log.debug("Error constructing HTTP request for HTTP adapter publish. No retries will be attempted.", e);
             return;
         }
 
-        printPublisherDiagnosticLog(eventContext, eventPayload, url,
+        printPublisherDiagnosticLog(eventProfileName, eventProfileUri, events, url,
                 HTTPAdapterConstants.LogConstants.ActionIDs.PUBLISH_EVENT, DiagnosticLog.ResultStatus.SUCCESS,
                 "Publishing event data to endpoint.");
 
         final long requestStartTime = System.currentTimeMillis();
-        final String correlationId = HTTPAdapterUtil.getCorrelationID(eventPayload);
 
         CompletableFuture<HttpResponse> future = clientManager.executeAsync(request);
 
         future.whenCompleteAsync((response, throwable) -> {
             try {
                 PrivilegedCarbonContext.startTenantFlow();
-                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(eventContext.getTenantDomain());
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain);
+
                 if (StringUtils.isNotEmpty(correlationId)) {
                     MDC.put(CORRELATION_ID_MDC, correlationId);
```

Comment:
Isn't it better to set the correlation id directly without checking existence likewise the tenant domain ?